### PR TITLE
GEOMESA-2973 Fix indexing for yearly z3 epoch

### PR DESCRIPTION
**Pull request by elahrvivaz**
* Does not fix z3 index versions < 5 (created with GeoMesa versions < 2.0.0)

Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

File: geomesa-z3/src/main/scala/org/locationtech/geomesa/curve/BinnedTime.scala [hunk 1]
```diff
@@ -150,7 +150,8 @@ object BinnedTime {
       case TimePeriod.Day   => ChronoUnit.DAYS.getDuration.toMillis
       case TimePeriod.Week  => ChronoUnit.WEEKS.getDuration.toMillis / 1000L
       case TimePeriod.Month => (ChronoUnit.DAYS.getDuration.toMillis / 1000L) * 31L
-      case TimePeriod.Year  => ChronoUnit.WEEKS.getDuration.toMinutes * 52L
+      // based on 365 days + 1 leap day, with a fudge factor of 10 minutes to account for leap seconds added each year
+      case TimePeriod.Year  => (ChronoUnit.DAYS.getDuration.toMinutes * 366L) + 10L
```
Comment (jnh5y) [inline geomesa-z3/src/main/scala/org/locationtech/geomesa/curve/BinnedTime.scala:154]: +1  Hopefully that'll hold us for awhile!

File: geomesa-z3/src/main/scala/org/locationtech/geomesa/curve/LegacyYearXZ3SFC.scala [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+package org.locationtech.geomesa.curve
+
+import java.time.temporal.ChronoUnit
+
+/**
+ * XZ3SFC with a legacy, incorrect max time value of 52 weeks. The max value is kept the same to ensure that
+ * index keys and query ranges are consistent. Any dates that exceed the original max time will be dropped into
+ * the last time bin, potentially degrading results for the last day or two of the year.
+ *
+ * @param g resolution level of the curve - i.e. how many times the space will be recursively quartered
```
Comment (jnh5y) [inline geomesa-z3/src/main/scala/org/locationtech/geomesa/curve/LegacyYearXZ3SFC.scala:18]: It is not quartered!  It is split into 8ths!;)

File: geomesa-z3/src/main/scala/org/locationtech/geomesa/curve/LegacyYearZ3SFC.scala [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+/**
+ * Z3SFC with a legacy, incorrect max time value of 52 weeks. The max value is kept the same to ensure that
+ * index keys and query ranges are consistent. Any dates that exceed the original max time will be dropped into
+ * the last time bin, potentially degrading results for the last day or two of the year.
+ *
+ * @param precision bits used per dimension - note all precisions must sum to less than 64
+ */
+@deprecated("Z3SFC", "3.2.0")
+class LegacyYearZ3SFC(precision: Int = 21) extends {
+  // need to use early instantiation here to prevent errors in creating parent class
```
Comment (jnh5y) [inline geomesa-z3/src/main/scala/org/locationtech/geomesa/curve/LegacyYearZ3SFC.scala:24]: Wow.  Tell me you hit this in a unit test and had to fix it:).

**Review by jnh5y — APPROVED** [review]
This is approved.  Will wait to merge until a backport is seen (in case it influences things).

Also, I think a note in the upgrade guide may be handy.

File: geomesa-z3/src/main/scala/org/locationtech/geomesa/curve/LegacyYearZ3SFC.scala (hunk 3 above)
Comment (elahrvivaz) [reply geomesa-z3/src/main/scala/org/locationtech/geomesa/curve/LegacyYearZ3SFC.scala:24]: indeed :)

File: geomesa-z3/src/main/scala/org/locationtech/geomesa/curve/LegacyYearXZ3SFC.scala (hunk 2 above)
Comment (elahrvivaz) [reply geomesa-z3/src/main/scala/org/locationtech/geomesa/curve/LegacyYearXZ3SFC.scala:18]: good point this was probably copied from the xz2 index

File: geomesa-index-api/src/test/scala/org/locationtech/geomesa/index/index/Z3IndexTest.scala [hunk 4]
```diff
@@ -55,6 +55,18 @@ class Z3IndexTest extends Specification with LazyLogging {
...
+
+        val lastDayFilter = ECQL.toFilter("bbox(geom,9,59,12,61) AND dtg during 2020-12-31T00:00:00.000Z/2020-12-31T23:59:59.999Z")
+
+        val lastDayResults =
+          SelfClosingIterator(ds.getFeatureReader(new Query("test", lastDayFilter), Transaction.AUTO_COMMIT)).toList
+
+        if (indices.indexOf("6") == -1) {
+          lastDayResults mustEqual Seq(features.last)
+        } else {
+          (lastDayResults mustEqual Seq(features.last)).pendingUntilFixed("requires strict bbox")
```
Comment (jnh5y) [inline geomesa-index-api/src/test/scala/org/locationtech/geomesa/index/index/Z3IndexTest.scala:67]: This returns data from 12/30, correct?

Comment (elahrvivaz) [reply geomesa-index-api/src/test/scala/org/locationtech/geomesa/index/index/Z3IndexTest.scala:67]: yeah, it returns the last 2 features

**Pull request merged**